### PR TITLE
Enable CI testing for C48-S2SWA-gefs on AWS and other CSPs

### LINKED ISSUE #3101: Enable CI testing for C48-S2SWA-gefs on AWS and other CSPs
### What new functionality do you need?

This is a prototype to Enable CI testing for C48-S2SWA-gefs on AWS and other CSPs

### What are the requirements for the new functionality?

Being able to initiate CI testing for C48-S2SWA-gefs at EMC G-W repo.

### Acceptance Criteria

Run successfully on AWS for C48-S2SWA-gefs case.

### Suggest a solution (optional)

This is test prototype, just put here for creating PR. 
Others do not need to work on this issue.
Wei

### PR DESCRIPTION
**Pull request by weihuang-jedi**
# Description

 Enabling CI testing on AWS for C48-S2SWA-gefs case.

  Resolves #3101
  
# Type of change
- [ ] Bug fix (fixes something broken)
- [x] New feature (adds functionality)
- [ ] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics

- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO (but may need to add with others at later time)
- Does this change require an update to any of the following submodules? NO (If YES, please add a link to any PRs that are pending.)
  - [ ] EMC verif-global 
  - [ ] GDAS 
  - [ ] GFS-utils 
  - [ ] GSI 
  - [ ] GSI-monitor 
  - [ ] GSI-utils 
  - [ ] UFS-utils 
  - [ ] UFS-weather-model 
  - [ ] wxflow 

# How has this been tested?

- Clone and build on AWS
- C48-S2SWA-gefs run on AWS

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have documented my code, including function, input, and output descriptions
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] This change is covered by an existing CI test or a new one has been added
- [ ] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [ ] I have made corresponding changes to the system documentation if necessary

### CODE REVIEW

**Comment by weihuang-jedi** [comment]
Please do review this draft PR.
This is a prototype test for CI testing on AWs for C48-S2SWA-gefs case.

File: ush/load_fv3gfs_modules.sh [hunk 1]
```diff
@@ -13,16 +13,32 @@ ulimit_s=$( ulimit -S -s )
 source "${HOMEgfs}/ush/detect_machine.sh"
 source "${HOMEgfs}/ush/module-setup.sh"
 
-# Source versions file for runtime
-source "${HOMEgfs}/versions/run.ver"
-
```
Comment (KateFriedman-NOAA) [inline ush/load_fv3gfs_modules.sh:18]: This should not be getting removed here. Please reinstate these lines that source `run.ver`. Will explain more in another comment in this file.

File: ush/load_fv3gfs_modules.sh [hunk 2]
```diff
@@ -13,16 +13,32 @@ ulimit_s=$( ulimit -S -s )
...
+    OS_NAME=$(grep -E '^ID=' /etc/os-release | sed -E 's/ID="?([^"]*)"?/\1/') || true
+    # Source versions file for runtime
+    source "${HOMEgfs}/versions/run.${MACHINE_ID}.${OS_NAME}.ver"
     module load "module_base.${MACHINE_ID}"
     ;;
+  "wcoss2" | "hera" | "orion" | "hercules" | "gaea" | "jet" | "s4")
+    # Source versions file for runtime
+    source "${HOMEgfs}/versions/run.${MACHINE_ID}.ver"
+    module load "module_base.${MACHINE_ID}"	    module load "module_base.${MACHINE_ID}"
+    ;;
```
Comment (KateFriedman-NOAA) [inline ush/load_fv3gfs_modules.sh:41]: Couple comments for these lines:

1. You do not need the lines to source `run.${MACHINE_ID}.ver`. The lines above to source `run.ver` should come back. When the link script is run the appropriate `run.${MACHINE}.ver` is copied or linked to become `run.ver`.
2. The logic to decide which `run.${MACHINE_ID}.${OS_NAME}.ver` should be moved to `sorc/link_workflow.sh`. This way, the correct run version file becomes `run.ver`. As discussed in comment 1.
3. I don't think any changes are needed to `ush/load_fv3gfs_modules.sh`. The changes I discussed in comments 1 and 2 should move all of these changes to the appropriate place.

Let me know if you have any questions about my comments. :)

File: versions/run.noaacloud.ver [hunk 3]
```diff
@@ -5,4 +5,8 @@ export spack_env=gsi-addon-env
 source "${HOMEgfs:-}/versions/spack.ver"
 export spack_mod_path="/contrib/spack-stack/spack-stack-${spack_stack_ver}/envs/gsi-addon-env/install/modulefiles/Core"
```
Comment (KateFriedman-NOAA) [inline versions/run.noaacloud.ver:6]: Do you still need this run version file (`run.noaacloud.ver`) if you have the other two now? If so, when will this one be used?

File: versions/spack.ver [hunk 4]
```diff
@@ -6,7 +6,7 @@ export jasper_ver=2.0.32
 export libpng_ver=1.6.37
 export zlib_ver=1.2.13
 export esmf_ver=8.5.0
-export fms_ver=2023.02.01
+export fms_ver=2023.04
```
Comment (KateFriedman-NOAA) [inline versions/spack.ver:9]: This will change the `fms_ver` for all platforms if changed in `spack.ver`. Is that the intent? If not, please add `export fms_ver=2023.04` to the noaacloud run version files.

File: versions/spack.ver [hunk 5]
```diff
@@ -23,7 +23,7 @@ export g2_ver=3.4.5
 export sp_ver=2.5.0
 export ip_ver=4.3.0
 export gsi_ncdiag_ver=1.1.2
-export g2tmpl_ver=1.10.2
+export g2tmpl_ver=1.13.0
```
Comment (KateFriedman-NOAA) [inline versions/spack.ver:26]: Same comment as with `fms_ver`.

File: sorc/ufs_noaacloud.intel.diff [hunk 6]
```diff
@@ -0,0 +1,10 @@
+--- ufs_noaacloud.intel.lua	2024-10-03 15:54:33.334583588 +0000
```
Comment (KateFriedman-NOAA) [inline sorc/ufs_noaacloud.intel.diff:1]: Will the need for this be resolved before this gets merged?

File: sorc/build_ufs.sh [hunk 7]
```diff
@@ -44,6 +44,14 @@ COMPILE_NR=0
...
 CLEAN_AFTER=NO
 
+#TODO temp patch for build update for noaacload in advance of updating ufs_module.fd repo for global-workflow building
+if [[ "${MACHINE_ID}" == "noaacloud" ]] ; then
+    patched=$(grep upp-addon-env modulefiles/ufs_noaacloud.intel.lua; echo $?)
+    if [[ ${patched} == "1" ]] ; then
+       patch -R modulefiles/ufs_noaacloud.intel.lua ../ufs_noaacloud.intel.diff
+    fi   
+fi
+
```
Comment (KateFriedman-NOAA) [inline sorc/build_ufs.sh:54]: Same question as one I also left in the newly added diff file. Can the need for this patch be resolved before merging this PR?

File: sorc/build_all.sh [hunk 8]
```diff
@@ -112,6 +112,9 @@ fi
 #------------------------------------
 # TODO: Commented out until components aligned for build
 #source ../versions/build.ver
+if [[ "${MACHINE_ID}" == "noaacloud" ]] ; then
+   source "../versions/build.${MACHINE_ID}.ver"
+fi
```
Comment (KateFriedman-NOAA) [inline sorc/build_all.sh:117]: At the moment we are allowing all components to build with their own `build.ver` files. Once all components are aligned to build with the same library versions we will uncomment line 114 and return to sourcing the global-workflow `build.ver` at build time, which overrides all build versions that the components. I'm guessing these lines were added to do a similar library version force on all of the components that are building on the cloud? Please confirm, thanks!

**Review by KateFriedman-NOAA — CHANGES_REQUESTED** [review]

File: sorc/ufs_noaacloud.intel.diff (hunk 6 above)
Comment (aerorahul) [reply sorc/ufs_noaacloud.intel.diff:1]: This diff should not be committed.  The ufs-weather-model needs to be able to be built on AWS without the need for a diff in the global-workflow

File: .github/workflows/pw_aws_ci.yaml [hunk 9]
```diff
@@ -114,7 +114,7 @@ jobs:
     - name: Build components
       run: |
         cd ${{ env.TEST_DIR }}/HOMEgfs/sorc
-        ./build_all.sh -j 8
+        ./build_all.sh -w -j 16
```
Comment (aerorahul) [inline .github/workflows/pw_aws_ci.yaml:117]: This will build the GEFS variant of the model (as is required for the GEFS test).  Doing so, will no longer test GFS tests.  
What/How do we plan to test GFS and GEFS simulateously?

File: sorc/build_ufs.sh (hunk 7 above)
Comment (aerorahul) [reply sorc/build_ufs.sh:54]: Work is needed in the ufs-weather-model to eliminate the application of the diff.  This is is a work around.

File: versions/run.noaacloud.centos.ver [hunk 10]
```diff
@@ -0,0 +1,12 @@
+export stack_intel_ver=2021.3.0
```
Comment (aerorahul) [inline versions/run.noaacloud.centos.ver:1]: If we are moving from CentOS to Rocky, why is this file required?

**Comment by weihuang-jedi** [comment]
This is a test trying to duplicate Terry's CI testing code, but trying to add c48gefs.
which did not work for now.
We need to switch to EPIC's new plan for CI testing, so close this one for now.
